src/GridGraph.flix

```diff
@@ -45,34 +53,32 @@ namespace Flixball/GridGraph {
             case None => None // no reachable goal
             case Some(end) =>
                 // construct the path list, end to start
-                def stepBack(c) = {
-                    let cLength = match MutMap.get(c, dists) {
-                        case None => ?unreachable
-                        case Some(l) => l
-                    };
-                    if (cLength == 1) None else
-
-                    let res = ref None;
-                    foreach(next <- possibleMoves(r, c)) {
-                        MutMap.get(next, dists) |>
-                            Option.foreach(d -> if (d == cLength - 1) res := Some(next) else ())
-                    };
-                    match deref res {
-                        case None => ?unreachable
-                        case Some(value) => Some(value)
-                    }
+                def stepBack(pos) = {
+                    let pathLength = MutMap.get(pos, dists) |> Utils/Option.unsafeGet;
+                    if (pathLength == 0) None else
+                    MutMap.get(pos, backtracking) |> Utils/Option.unsafeGet |> Some
                 };
-                let revPathList = List.unfold(c ->  stepBack(c) |> Option.map(next -> (next, next)), end);
-                Some(List.reverse(end :: revPathList))
+                let revPathList = List.unfold(c -> stepBack(c) |> Option.map(match (prev, move) -> (move, prev)), end);
+                Some(List.reverse(revPathList))
         }
     }}
 
-    def possibleMoves(r: Region[r], c: Coordinates): Iterator[Coordinates, r] \ r =
-        let x = fst(c);
-        let y = snd(c);
-        let coordinates = (x, y+1) :: (x, y-1) :: (x+1, y) :: (x-1, y) :: Nil;
-        List.iterator(r, coordinates)
-
+    def possibleMoves(r: Region[r], p: Position): Iterator[(Position, Move), r] \ Write(r) =
```

Comment:
Let's add a documentation comment. I assume the pair represents the move and the new position after the move.

Reply:
Why do we use an iterator rather than a list, since it's just 3 elements?

Reply:
No reason, really. It's just used in a `foreach` so it has to become an iterator anyway. Should probably be changed if it becomes `pub`